Added multi-lane support to BasecallsConverter

jacarey opened the pull request:
### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

Commit:
Added multi-lane support to BasecallsConverter

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -69,7 +69,7 @@
      * @param basecallsDir                 Where to read basecalls from.
      * @param barcodesDir                  Where to read barcodes from (optional; use basecallsDir if not specified).
-     * @param lane                         What lane to process.
+     * @param lanes                         What lanes to process.
```
```suggestion
     * @param lanes                        What lanes to process.
```

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -452,7 +453,7 @@ private void populateWritersFromLibraryParams() {
         final Map<String, String> params = new LinkedHashMap<>();
 
-        String platformUnit = RUN_BARCODE + "." + LANE;
+        String platformUnit = RUN_BARCODE + "." + LANE.stream().map(Object::toString).collect(Collectors.joining(","));
```
I wonder if it would be more correct to put each lane into a separate read group?  That's effectively what you'd get today if you ran per-lane and then merged the resulting BAMs.  Any idea how hard that would be to implement?

tfenne commented on src/main/java/picard/illumina/SortedBasecallsConverter.java:
```diff
@@ -152,28 +153,26 @@ public void run() {
-                if (includeNonPfReads || cluster.isPf()) {
-                    addRecord(cluster.getMatchedBarcode(), converter.convertClusterToOutputRecord(cluster));
+            final List<SortedRecordToWriterPump> writerList = new ArrayList<>();
+            for(IlluminaDataProviderFactory laneFactory : laneFactories) {
```
```suggestion
            for (IlluminaDataProviderFactory laneFactory : laneFactories) {
```

Commit:
[refactor] Large refactor of IlluminaBaseCallsConverter (#1626)

* Normalize data provider entry points
  Unified basecalls converter
* First pass at BasecallsConverterBuilder
* Pass two for BasecallsConverterBuilder
* Fix bug in tile fetch for multi tile data.
* Update to BasecallsConverter abstract class.
* Added UnsortedBasecallsConverter.
* Speed up encoding of Casava read names.
* Attempt to make things much more efficient when not sorting by bypassing the conversion from ClusterData into FastqRecordsForCluster.
BclReader now reads from it's streams into a small buffer to speed things up.
* Used pre-generated lookups for decoding BCL bases and quals.
* Review changes.
Added ClusterDataCodec so that SortedBasecallsConverter can use ClusterWriter instead of FastqRecordWriter for SortingCollections.
* Added basic write threading to sorted converter.
* Added basic write threading to unsorted converter.
* Fix to ensure that IOUtil.openFileForWriting() will respect CommandLineProgram.COMPRESSION_LEVEL when writing gzipped files.
* Fix bug in thread count setting.
* Small performance tweaks.
* Change tests using deprecated `MAX_READS_IN_RAM_PER_TILE`.
Adjust error checking on executors to avoid potential race conditions.
* Review changes.
Added a lot of unsorted test data for IlluminaBasecallsToSam.
* Remove unsorted test data.
Fix for potential race condition.

Co-authored-by: Jay Carey <jay@fulcrumgenomics.com>
Co-authored-by: tfenne <tfenne@tfenne.com>
Co-authored-by: George Grant <ggrant@broadinstitute.org>
Co-authored-by: Nils Homer <nh13@users.noreply.github.com>
Co-authored-by: kachulis <39926576+kachulis@users.noreply.github.com>

tfenne reviewed (changes requested):
Looks good to me - I just want to make sure we talk about read groups in IBTS before we open the public PR.

cmnbroad committed:
Include common (engine-level) args in help output. (#1580)

Co-authored-by: George Grant <ggrant@broadinstitute.org>

tfenne commented on src/main/java/picard/illumina/ClusterDataToSamConverter.java:
```diff
@@ -199,8 +200,8 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
 
-        if (this.readGroupId != null) {
-            sam.setAttribute(SAMTag.RG.name(), readGroupId);
+        if (this.laneToReadGroup.containsKey(lane)) {
+            sam.setAttribute(SAMTag.RG.name(), laneToReadGroup.get(lane));
```
```suggestion
        final String rgId = this.laneToReadGroup.get(lane);
        if (rgId != null) sam.setAttribute(SAMTag.RG.name(), rgId);
```
Avoid hitting the map twice.

tfenne reviewed (approved):
One small suggestion otherwise looks good to me.

lbergelson committed:
Update htsjdk 2.23.0 -> 2.24.0 (#1641)

Commit:
Merge branch 'master' into jc_multilane_support
